File: auth-service/src/utils/tracing.ts
import { NodeSDK } from '@opentelemetry/sdk-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'; // Using HTTP for wider compatibility, can be GRPC
import { Resource } from '@opentelemetry/resources';
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions';
import { SimpleSpanProcessor, ConsoleSpanExporter, BatchSpanProcessor } from '@opentelemetry/sdk-trace-node';
import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
import { PinoInstrumentation } from '@opentelemetry/instrumentation-pino'; // If using Pino logger
import { MongoDBInstrumentation } from '@opentelemetry/instrumentation-mongodb'; // If using MongoDB
import { RedisInstrumentation } from '@opentelemetry/instrumentation-redis-4'; // For redis ioredis v4

import { logger } from './logger'; // Assuming your logger is here

export const initTracer = (serviceName: string, jaegerEndpoint: string, nodeEnv: string) => {
  const traceExporter = new OTLPTraceExporter({
    url: jaegerEndpoint, // e.g., http://localhost:4318/v1/traces for OTLP HTTP
  });

  const spanProcessor = nodeEnv === 'production'
    ? new BatchSpanProcessor(traceExporter)
    : new SimpleSpanProcessor(new ConsoleSpanExporter()); // Use ConsoleSpanExporter for dev/debug

  const sdk = new NodeSDK({
    resource: new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: serviceName,
      [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: nodeEnv,
    }),
    spanProcessor: spanProcessor,
    // traceExporter, // Not needed if using BatchSpanProcessor explicitly
    instrumentations: [
      new HttpInstrumentation(),
      new ExpressInstrumentation(),
      new PinoInstrumentation({ // Assuming Pino logger is used, configure as needed
        logHook: (_span, record) => {
            record['resource.service.name'] = serviceName;
        },
      }),
      new MongoDBInstrumentation({
        // Enable enhanced database reporting elsewhere in your application
        enhancedDatabaseReporting: true,
      }),
      new RedisInstrumentation(), // For ioredis v4+
    ],
  });

  try {
    sdk.start();
    logger.info(`OpenTelemetry Tracing initialized for service: ${serviceName}, environment: ${nodeEnv}`);
    logger.info(`Jaeger exporter configured with endpoint: ${jaegerEndpoint}`);
    if (nodeEnv !== 'production') {
      logger.info('ConsoleSpanExporter is active for development/debugging.');
    }
  } catch (error) {
    logger.error('Error initializing OpenTelemetry Tracing:', error);
  }

  // Graceful shutdown
  process.on('SIGTERM', () => {
    sdk.shutdown()
      .then(() => logger.info('OpenTelemetry Tracing terminated'))
      .catch((error) => logger.error('Error terminating OpenTelemetry Tracing', error))
      .finally(() => process.exit(0));
  });

  process.on('SIGINT', () => {
    sdk.shutdown()
      .then(() => logger.info('OpenTelemetry Tracing terminated via SIGINT'))
      .catch((error) => logger.error('Error terminating OpenTelemetry Tracing via SIGINT', error))
      .finally(() => process.exit(0));
  });

  return sdk;
};

File: auth-service-py/app/utils/tracing.py
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, ConsoleSpanExporter # For debugging
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter as OTLPHttpSpanExporter
# from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter as OTLPGrpcSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME as RESOURCE_SERVICE_NAME

from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.instrumentation.redis import RedisInstrumentor
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor


from app.core.config import settings
from app.utils.logger import logger

def init_tracer(service_name: str = settings.OTEL_SERVICE_NAME, jaeger_endpoint: str = settings.JAEGER_ENDPOINT):
    """
    Initializes OpenTelemetry tracing.
    """
    if not settings.ENABLE_TRACING:
        logger.info("OpenTelemetry tracing is disabled.")
        return

    try:
        resource = Resource(attributes={
            RESOURCE_SERVICE_NAME: service_name,
            "environment": settings.ENVIRONMENT,
            # Add other common resource attributes here
        })

        provider = TracerProvider(resource=resource)

        # OTLP Exporter (HTTP by default, can switch to gRPC)
        otlp_exporter = OTLPHttpSpanExporter(endpoint=jaeger_endpoint)
        # otlp_exporter = OTLPGrpcSpanExporter(endpoint="localhost:4317") # Example for gRPC

        processor = BatchSpanProcessor(otlp_exporter)
        provider.add_span_processor(processor)

        # Console exporter for debugging (optional)
        if settings.ENVIRONMENT == "development":
            console_exporter = ConsoleSpanExporter()
            console_processor = BatchSpanProcessor(console_exporter)
            provider.add_span_processor(console_processor)
            logger.debug("ConsoleSpanExporter enabled for OpenTelemetry tracing.")

        trace.set_tracer_provider(provider)
        logger.info(f"OpenTelemetry tracing initialized for service: {service_name}, exporting to {jaeger_endpoint}")

    except Exception as e:
        logger.error(f"Failed to initialize OpenTelemetry tracer: {e}", exc_info=True)


def instrument_fastapi_app(app):
    """Instruments a FastAPI application and related libraries."""
    if not settings.ENABLE_TRACING:
        return

    try:
        FastAPIInstrumentor.instrument_app(app, tracer_provider=trace.get_tracer_provider())

        # Assuming 'engine' is your SQLAlchemy engine instance
        # from app.db.database import engine
        # SQLAlchemyInstrumentor().instrument(engine=engine) # Instrument after engine creation

        RedisInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())
        HTTPXClientInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())

        logger.info("FastAPI application and common libraries instrumented for OpenTelemetry.")
    except Exception as e:
        logger.error(f"Failed to instrument FastAPI app for OpenTelemetry: {e}", exc_info=True)

# Example of getting a tracer
# tracer = trace.get_tracer(__name__)
# with tracer.start_as_current_span("my-custom-span"):
#     # do some work
#     pass
